python/plugins/processing/algs/qgis/Difference.py

```diff
@@ -84,23 +84,34 @@ def processAlgorithm(self, progress):
             diff_geom = QgsGeometry(geom)
             attrs = inFeatA.attributes()
             intersections = index.intersects(geom.boundingBox())
+            
+            unionGeom = QgsGeometry()
+            geomList = []
+
             for i in intersections:
                 request = QgsFeatureRequest().setFilterFid(i)
                 inFeatB = layerB.getFeatures(request).next()
-                tmpGeom = QgsGeometry(inFeatB.geometry())
-                if diff_geom.intersects(tmpGeom):
-                    diff_geom = QgsGeometry(diff_geom.difference(tmpGeom))
-                    if diff_geom.isGeosEmpty():
-                        ProcessingLog.addToLog(ProcessingLog.LOG_INFO,
-                                               self.tr('Feature with NULL geometry found.'))
-                    if not diff_geom.isGeosValid():
-                        if ignoreInvalid:
-                            ProcessingLog.addToLog(ProcessingLog.LOG_ERROR,
-                                                   self.tr('GEOS geoprocessing error: One or more input features have invalid geometry.'))
-                            add = False
-                        else:
-                            raise GeoAlgorithmExecutionException(self.tr('Features with invalid geometries found. Please fix these errors or specify the "Ignore invalid input features" flag'))
-                        break
+                
+                if inFeatB.geometry().isGeosEmpty():
+                    ProcessingLog.addToLog(ProcessingLog.LOG_INFO,
+                                           self.tr('Feature with NULL geometry found.'))
+
+                if not inFeatB.geometry().isGeosValid():
+                    if ignoreInvalid:
+                        ProcessingLog.addToLog(ProcessingLog.LOG_ERROR,
+                                               self.tr('GEOS geoprocessing error: One or more input features have invalid geometry.'))
+                        add = False
+                    else:
+                        raise GeoAlgorithmExecutionException(self.tr('Features with invalid geometries found. Please fix these errors or specify the "Ignore invalid input features" flag'))
+                    break
+              
+                if (inFeatB.geometry().intersects(geom)):
+                    geomList.append(QgsGeometry(inFeatB.geometry()))
```

Comment:
No need gor the extra QgsGeometry
Out of curiosity, did you try using unaryUnion here? Does this also perform slow?

Reply:
Not sure I understund you correct but I'm collecting all intersecting geometry copies in list because I faced some segmentation fault when I tryed union them one by one in a loop. I got impression that it has something to do references of feature geometries which points to nowhere when loop scope ends. However this approach seems to do the job.

Reply:
Actually @m-kuhn is mistaken here - you do need the extra QgsGeometry to force a full copy. It's not needed for 3.0 anymore, but given that this is targeted to 2.18 it will be needed here.

On that note - please also forward port these changes to master. It's getting extremely difficult to track the amount of changes applied to processing algs in 2.18 which haven't been added to master.

Reply:
Ah right, sorry, I missed that this targets 2.18